xrootd/client/session.go

```diff
@@ -0,0 +1,252 @@
+	client           *Client
+}
+
+func newSession(ctx context.Context, address string, username string, client *Client) (*session, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	var d net.Dialer
+	conn, err := d.DialContext(ctx, "tcp", address)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	session := &session{cancel: cancel, conn: conn, mux: mux.New(), requests: make(map[xrdproto.StreamID][]byte), client: client}
+
+	go session.consume(ctx)
+
+	if err := session.handshake(ctx); err != nil {
+		session.Close()
+		return nil, err
+	}
+
+	securityInfo, err := session.Login(ctx, username, "")
+	if err != nil {
+		session.Close()
+		return nil, err
+	}
+
+	if len(securityInfo.SecurityInformation) > 0 {
+		err = session.auth(ctx, securityInfo.SecurityInformation)
+		if err != nil {
+			session.Close()
+			return nil, err
+		}
+	}
+
+	protocolInfo, err := session.Protocol(ctx)
+	if err != nil {
+		session.Close()
+		return nil, err
+	}
+
+	session.signRequirements = signing.New(protocolInfo.SecurityLevel, protocolInfo.SecurityOverrides)
+
+	return session, nil
+}
+
+// Close closes the connection. Any blocked operation will be unblocked and return error.
+func (session *session) Close() error {
+	session.cancel()
+
+	session.mux.Close()
+	return session.conn.Close()
+}
+
+func (session *session) consume(ctx context.Context) {
+	var header xrdproto.ResponseHeader
+	var headerBytes = make([]byte, xrdproto.ResponseHeaderLength)
+
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: Should wait for active requests to be completed?
+			return
+		default:
+			if _, err := io.ReadFull(session.conn, headerBytes); err != nil {
+				if ctx.Err() != nil {
+					// something happened to the context.
+					// ignore this error.
+					continue
+				}
+				panic(err)
+				// TODO: handle EOF by redirection as specified at http://xrootd.org/doc/dev45/XRdv310.pdf, page 11
+			}
+
+			if err := xrdproto.Unmarshal(headerBytes, &header); err != nil {
+				if ctx.Err() != nil {
+					// something happened to the context.
+					// ignore this error.
+					continue
+				}
+				panic(err)
+				// TODO: should redirect in case if is not possible to decode a header as well?
+			}
+
+			resp := mux.ServerResponse{Data: make([]byte, header.DataLength)}
+			if _, err := io.ReadFull(session.conn, resp.Data); err != nil {
+				if ctx.Err() != nil {
+					// something happened to the context.
+					// ignore this error.
+					continue
+				}
+				resp.Err = err
+			}
+
+			switch header.Status {
+			case xrdproto.Error:
+				resp.Err = header.Error(resp.Data)
+			case xrdproto.Wait:
+				go func(streamID xrdproto.StreamID) {
+					// TODO: fetch sleep value from response.
+					time.Sleep(time.Second * 5)
+					if _, err := session.conn.Write(session.requests[streamID]); err != nil {
+						log.Printf("received error: %v", err)
+					}
+				}(header.StreamID)
+				continue
+			case xrdproto.Redirect:
+				port := binary.BigEndian.Uint32(resp.Data)
```

Comment:
this whole port+opaque+token extraction business should probably be lifted as a function, which could then be properly tested for a bunch of cases.
```go
func parseRedirect(raw []byte) (host string, port int, opaque, token string, err error) {
    ...
}
```

or, directly inside `internal/mux`:
```go
func ParseRedirection(raw []byte) (*Redirection, error) {
    ...
}
```